sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala

```diff
@@ -200,15 +200,19 @@ class FrameLessOffsetWindowFunctionFrame(
 
   override def prepare(rows: ExternalAppendOnlyUnsafeRowArray): Unit = {
     resetStates(rows)
-    if (ignoreNulls) {
-      findNextRowWithNonNullInput()
+    if (Math.abs(offset) > rows.length) {
+      inputIndex = offset
     } else {
-      // drain the first few rows if offset is larger than zero
-      while (inputIndex < offset) {
-        if (inputIterator.hasNext) inputIterator.next()
-        inputIndex += 1
+      if (ignoreNulls) {
+        findNextRowWithNonNullInput()
+      } else {
+        // drain the first few rows if offset is larger than zero
+        while (inputIndex < offset) {
+          if (inputIterator.hasNext) inputIterator.next()
+          inputIndex += 1
+        }
+        inputIndex = offset
       }
-      inputIndex = offset
```

Comment:
The root cause is `findNextRowWithNonNullInput` required `assert(offset <= input.length)`, I think we can simplify the logic here!

```
    if (offset > rows.length) {
      fillDefaultValue(EmptyRow)
    } else {
      resetStates(rows)
      if (ignoreNulls) {
        findNextRowWithNonNullInput()
      } else {
        // drain the first few rows if offset is larger than zero
        while (inputIndex < offset) {
          if (inputIterator.hasNext) inputIterator.next()
          inputIndex += 1
        }
        inputIndex = offset
      }
    }
```

Reply:
yes, I considered this but it does not work. The actual projection (or default value) is created in `doWrite` so `prepare` needs to initialize variables such as `input`, `inputIndex` etc.

Reply:
I tested with the code I mentioned, It works good!

```
  override def prepare(rows: ExternalAppendOnlyUnsafeRowArray): Unit = {
    resetStates(rows)
    if (offset > rows.length) {
      fillDefaultValue(EmptyRow)
    } else {
      resetStates(rows)
      if (ignoreNulls) {
        findNextRowWithNonNullInput()
      } else {
        // drain the first few rows if offset is larger than zero
        while (inputIndex < offset) {
          if (inputIterator.hasNext) inputIterator.next()
          inputIndex += 1
        }
        inputIndex = offset
      }
    }
  }
```

Reply:
yes, it works when `ignoreNulls == true` but breaks when `ignoreNulls == false`

Reply:
Modified PR to localize change for `ignoreNulls` case only, please take a look